### PR TITLE
Remote Debugging Nautobot and Celery-Workers

### PR DESCRIPTION
**Pull request by jifox**
# Closes: #1273
# What's Changed

Documentation on how to configure VS Code for remote debugging **Nautobot** and **Celery_Worker** running in container

# TODO

- [ ] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: changes/3081.added [hunk 1]
```diff
@@ -0,0 +1 @@
+Add section "VScode Remote Debugging Configuration" to development chapter in documentation
```
Comment (gsnider2195) [inline changes/3081.added:1]:
```suggestion
Added section "VScode Remote Debugging Configuration" to development chapter in documentation.
```

File: nautobot/docs/development/docker-compose-advanced-use-cases.md [hunk 2]
```diff
@@ -165,3 +165,132 @@ $ docker-compose -f docker-compose.yml -f docker-compose.debug.yml up
 - The **Select the container to attach VS Code** input field provides a list of running containers.
 - Click on `development_nautobot_1` to use VS Code inside the container. The `devcontainer` will startup now.
 - As a last step open the folder `/opt/nautobot` in VS Code.
+
+## VScode Remote Debugging Configuration
+
+Using the Remote-Attach functionality of VS Code debugger is an alternative to debugging in  development container. This allows a local VS Code instance to connect to a remote container and debug the code running in the container the same way as when debugging locally.
```
Comment (gsnider2195) [inline nautobot/docs/development/docker-compose-advanced-use-cases.md:171]:
```suggestion
Using the Remote-Attach functionality of VS Code debugger is an alternative to debugging in a development container. This allows a local VS Code instance to connect to a remote container and debug the code running in the container the same way as when debugging locally.
```

File: nautobot/docs/development/docker-compose-advanced-use-cases.md [hunk 3]
```diff
@@ -165,3 +165,132 @@ $ docker-compose -f docker-compose.yml -f docker-compose.debug.yml up
...
+            "port": 6898
+          },
+          "pathMappings": [
+            {
+              "localRoot": "${workspaceFolder}",
+              "remoteRoot": "/source"
+            }
+          ],
+          "django": true
+        }
```
Comment (gsnider2195) [inline nautobot/docs/development/docker-compose-advanced-use-cases.md:289]: These can be added to the `launch` section in `nautobot.code-workspace`

**Comment by gsnider2195** [comment]
Instead of modifying the dockerfile I was able to get this working by doing a `poetry add --dev debugpy` and then changing the command here to `python -m debugpy --listen 0.0.0.0:6899 -m manage runserver 0.0.0.0:8080 --insecure`. I recommend we add debugpy to the poetry dev dependencies and move the vscode debug config to the `nautobot.code-workspace` file. Then this documentation can be condensed to just the `docker-compose.override.yml` changes.

File: nautobot/docs/development/docker-compose-advanced-use-cases.md [hunk 4]
```diff
@@ -165,3 +165,132 @@ $ docker-compose -f docker-compose.yml -f docker-compose.debug.yml up
...
+
+    # Add vscode remote debug support
+    RUN  pip install --no-cache-dir --upgrade debugpy  -t /tmp
+    ```
+
+2. **Configure `invoke.yml` to use the `docker-compose.override.yml` file**
+
+    This allows to modify container settings without touching the original `docker-compose.yml` file.
+
+    Copy the file `invoke.yml.example` to `invoke.yml` and configure at least the `compose_dir` according your project location.
```
Comment (gsnider2195) [inline nautobot/docs/development/docker-compose-advanced-use-cases.md:203]:
```suggestion
    Copy the file `invoke.yml.example` to `invoke.yml` and configure at least the `compose_dir` according your project location.
    
    See the [docker compose override](development/docker-compose-advanced-use-cases.md#docker-compose-overrides) documentation for more details.
```

Or something similar.

**Comment by gsnider2195** [comment]
Thank you for taking the time to figure this out! I spent a lot of time trying to figure this out myself and didn't make any progress.

File: changes/3081.added [hunk 5]
```diff
@@ -0,0 +1 @@
+Added section "VScode Remote Debugging Configuration" to development chapter in documentation.
```
Comment (jathanism) [inline changes/3081.added:1]:
```suggestion
Added section "VS Code Remote Debugging Configuration" to development chapter in documentation.
```

File: nautobot/docs/development/docker-compose-advanced-use-cases.md [hunk 6]
```diff
@@ -165,3 +165,113 @@ $ docker-compose -f docker-compose.yml -f docker-compose.debug.yml up
 - The **Select the container to attach VS Code** input field provides a list of running containers.
 - Click on `development_nautobot_1` to use VS Code inside the container. The `devcontainer` will startup now.
 - As a last step open the folder `/opt/nautobot` in VS Code.
+
+## VScode Remote Debugging Configuration
```
Comment (jathanism) [inline nautobot/docs/development/docker-compose-advanced-use-cases.md:169]:
```suggestion
## VS Code Remote Debugging Configuration
```

**Review by jathanism — COMMENTED** [review]
Just nitpicking for consistency in referencing the app "VS Code" as compared to the existing documentation.

**Comment by jifox** [comment]
@gsnider2195

> Thank you for taking the time to figure this out!

Your'e welcome.

If you integrate `debugpy` as a dev requirement, that would be best.

I'm not sure if it's best to integrate the new start commands in `Dockerfile` too because that's changing all dependent containers. I would suggest adding a docker-compose.override.yml.example with this startup commands as described. This will be most flexible and has least impact to containers using the dev containers as a basis.

File: changes/3081.added [hunk 7]
```diff
@@ -0,0 +1 @@
+Added section "VS Code Remote Debugging Configuration" to development chapter in documentation.
```
Comment (gsnider2195) [inline changes/3081.added:1]: This filename should be `1273.added`

File: nautobot/docs/development/docker-compose-advanced-use-cases.md (hunk 3 above)
Comment (jifox) [reply nautobot/docs/development/docker-compose-advanced-use-cases.md]: Done

**Comment by jifox** [comment]
How to proceed? Modifying Dockerfile or add docker-compose.override.yml?

**Comment by gsnider2195** [comment]
> How to proceed? Modifying Dockerfile or add docker-compose.override.yml?

Since the dockerfile is already running `poetry install --extras all` then it's more concise to add any new dependencies as poetry dev dependencies. I'd like to see if vscode can leverage the celery remote debugger [rdb](https://docs.celeryq.dev/en/stable/reference/celery.contrib.rdb.html) before we add a new dev dependency. I probably won't be able to get to that this week so if that's something you'd like to look into let me know.

**Comment by gsnider2195** [comment]
The other option is to pip install the dependency in the docker compose override file. Something like `pip install debugpy && python -m debugpy --listen 0.0.0.0:6899 -m manage runserver 0.0.0.0:8080 --insecure`

**Comment by bryanculver** [comment]
@jifox Please pull down the latest commit and validate the changes still work for you locally.

**Comment by jifox** [comment]
@bryanculver Thanks for tweaks, It works like a charm. I tested Nautobot and Nautobot-Celery debugging.

**Review by bryanculver — APPROVED** [review]

**Pull request merged**